opentelemetry-sdk/src/opentelemetry/sdk/metrics/export/exemplars.py

```diff
@@ -0,0 +1,349 @@
+            self._span_id,
+            self._trace_id,
+        )
+
+    @property
+    def value(self):
+        """The current value of the Exemplar point"""
+        return self._value
+
+    @property
+    def timestamp(self):
+        """The time that this Exemplar's value was recorded"""
+        return self._timestamp
+
+    @property
+    def span_id(self):
+        """The span ID of the context when the exemplar was recorded"""
+        return self._span_id
+
+    @property
+    def trace_id(self):
+        """The trace ID of the context when the exemplar was recorded"""
+        return self._trace_id
+
+    @property
+    def dropped_labels(self):
+        """Labels that were dropped by the aggregator but still passed by the user"""
+        return self._dropped_labels
+
+    @property
+    def sample_count(self):
+        """For statistical exemplars, how many measurements a single exemplar represents"""
+        return self._sample_count
+
+    def set_sample_count(self, count):
+        self._sample_count = count
+
+
+class ExemplarSampler:
+    """
+        Abstract class to sample exemplars through a stream of incoming measurements
+    """
+
+    def __init__(self, k, statistical=False):
+        self._k = k
+        self._statistical = statistical
+        self._sample_set = list()
+
+    @abc.abstractmethod
+    def sample(self, exemplar, **kwargs):
+        """
+        Given an exemplar, determine if it should be sampled or not
+        """
+
+    @property
+    @abc.abstractmethod
+    def sample_set(self):
+        """
+        Return the list of exemplars that have been sampled
+        """
+
+    @abc.abstractmethod
+    def merge(self, set1, set2):
+        """
+        Given two lists of sampled exemplars, merge them while maintaining the invariants specified by this sampler
+        """
+
+    @abc.abstractmethod
+    def reset(self):
+        """
+        Reset the sampler
+        """
+
+
+class RandomExemplarSampler(ExemplarSampler):
+    """
+        Randomly sample a set of k exemplars from a stream. Each measurement in the stream
+        will have an equal chance of being sampled.
+
+        If RandomExemplarSampler` is specified to be statistical, it will add a sample_count to every exemplar it records.
+        This value will be equal to the number of measurements recorded per every exemplar measured - all exemplars will have the same sample_count value.
+    """
+
+    def __init__(self, k, statistical=False):
+        super().__init__(k, statistical=statistical)
+        self.rand_count = 0
+
+    def sample(self, exemplar, **kwargs):
+        self.rand_count += 1
+
+        if len(self.sample_set) < self._k:
+            self.sample_set.append(exemplar)
+            return
+
+        replace_index = random.randint(0, self.rand_count - 1)
+
+        if replace_index < self._k:
+            self.sample_set[replace_index] = exemplar
+
+    def merge(self, set1, set2):
+        combined = set1 + set2
```

Comment:
Looking at the algorithm Josh linked, I don't think merging like this would give a uniform sample. In the extreme case, imagine `k = 1` and you have two samplers; sampler1's `rand_count = 1` (population size of 1)  and sampler2 finishes with `rand_count = 10`. sampler1's single value would be weighted much more strongly than any of the values sampler2 saw.

E.g. sampler1 see value `[0]`, sampler2 sees values `[1, 2, 3, ..., 10]` but keeps only k = 1 sample, say, `[6]`. Randomly sampling k = 1 values from `[0, 6]` would give `[0]` 50% of the time, even though it was one of 11 values sampled in the whole population.

I think there is a similar issue when `arg_count > k`. The wiki page says each item is sampled with probability `k/n`; if the samplers being merged saw different `n`, then you would end up giving more weight to the sampler with smaller `n`.

Reply:
as discussed offline, just keeping the second set's exemplars with that assumption based on the current SDK implementation. Might need to change in the future but for now that keeps this much simpler